### ISSUE TITLE
--value='end' for non-periodic queries ignores P directives after last posting

### ISSUE THREAD

**Issue by aragaer** [simonmichael/hledger#1405]
Sample journal:
```
P 2020-11-20 commodity 1

2020-11-20
 (assets)  1 commodity

P 2020-11-22 commodity 2
```
Now the following two commands show different result:
`hledger bal -f /tmp/test.journal --value='end,' --cumulative -e 2021` shows 1, while `hledger bal -f /tmp/test.journal --value='end,' --cumulative -e 2021 -Y` shows 2. Adding an unrelated posting fixes that:
```
2020-11-22
 unrelated  0
```
I have found this accidentally by comparing yearly results to "current" value. This difference in behavior seems a bit non-intuitive.

**Comment by simonmichael** [comment]
Thanks for the report. I agree there might be an issue here. --debug can give more insight. I thought -V was equivalent to --value=end, but apparently not always:
```journal
P 2020-11-20 A 1B

2020-11-20
 (assets)  1A

P 2020-11-22 A 2B
```
```shell
$ hledger bal -e 2021 --value=end --debug=1 2>&1 | grep 'valuation date'
valuation date: 2020-11-20
$ hledger bal -e 2021 -V --debug=1 2>&1 | grep 'valuation date'
valuation date: 2020-12-31
```

https://hledger.org/hledger.html#valuation-date says:
> For single period reports, if an explicit report end date is specified, that will be used as the valuation date; otherwise the valuation date is "today".

-V used the report end date, but --value=end used the journal end date (date of last transaction).
https://hledger.org/hledger.html#value-flexible-valuation says:
> --value=end : Convert amounts to their value in the default valuation commodity, using market prices on the last day of the report period (or if unspecified, the journal's end date); or in multiperiod reports, market prices on the last day of each subperiod.

Last day of the report period was specified, so I think it should have used that, like -V. I'm only slight concerned that it was set this way for a reason, so we should check the commit history. Another part of the doc seems to acknowledge the difference in behaviour:

https://hledger.org/hledger.html#effect-of-valuation-on-reports > balance > balance changes says for -V:

> value at report end or today of sums of postings

and for --value=end:

> value at report or journal end of sums of postings

**Comment by aragaer** [comment]
The only thing I might propose is that P directive is considered a journal entry when we're looking for "the end of journal/period".

**Comment by Xitian9** [comment]
I agree with @aragaer that including P directives in the journal span should happen, but perhaps only for the journal end date. I imagine people might put some historical prices into their journals, and these should probably not effect display at all.

**Comment by simonmichael** [comment]
#1416 has been merged, so in theory P directives now contribute to the default report end date, but I'll note that it does not affect my test above (with --value=end it's still picking 2020-11-20 as valuation date). So perhaps a little more needed, if I'm not confused.

**Comment by Xitian9** [comment]
Yes, valuation dates are not quite working properly yet. Working on that now.

**Comment by Xitian9** [comment]
Here's a question: is it worth maintaining this difference between `-V` and `--value=end`? The currently only differ in the following way:

If running a single-period report with no explicit end date:
   - with `--empty` option
      - `-V` will value at `today`
      - `--value=end` will value at the journal end date, i.e. the most recent posting, transaction, or market price date
   - without `--empty` option
      - `-V` will value at `today`
      - `--value=end` will value at the most recent posting date for the included report
      
I'm proposing some changes addressing #1396; these will make everything behave as if `--empty` were specified. However, this still leaves the question of whether we should keep the `-V`/`--value=end` difference. The only functional difference at this point is if you have future transactions or market prices in your journal, in which case `--value=end` will see them but `-V` won't. This seems like a rather niche use case to have error-prone logic to handle, so I would propose just having `-V` for single period reports behave identically to `--value=end`. Thoughts?

**Comment by Xitian9** [comment]
Past instances in which this behaviour was noted: #453, #683.

**Comment by simonmichael** [comment]
@Xitian9 I haven't yet managed to answer in detail but FWIW I intended -V to be equivalent to --value=end in a multiperiod report, and equivalent to --value=now in a single period report. 

A piece of doc related to this topic is https://hledger.org/hledger.html#valuation-date.

**Comment by Xitian9** [comment]
I can definitely see the reasoning for having it as --value=now before the inclusion of #1416: any price directives after the last posting would not be seen with --value=end, even if they were in the past, but --value=now would see them. However, --value=end will now see these price directives after #1416 as well.

Maybe others think differently, but if I ask for a single balance report for some accounts in 2013, I would expect them to be valued at the end of 2013, not the current date. If I specifically wanted them at the current date, calling --value=now seems more explicit and clear. This is especially true since multi balance reports already act this way.

**Comment by simonmichael** [comment]
That sounds good but I'm confused. Is there a small example or an existing or proposed test that can demonstrate what you're proposing ?

**Comment by Xitian9** [comment]
pastvaluation.journal
```
2000-01-01  Test
    a  1 A
    b

P 2001-01-01 A 2 B
```

Output for v1.20.1
```
$ hledger -f pastvaluation.journal bal a -N -V
                 2 B  a
$ hledger -f pastvaluation.journal bal a -N --value=end
                 1 A  a
```

Proposed output
```
$ hledger -f pastvaluation.journal bal a -N -V
                 2 B  a
$ hledger -f pastvaluation.journal bal a -N --value=end
                 2 B  a
```

futurevaluation.journal
```
P 2000-01-01 A 1 B
P 2050-01-01 A 2 B

2100-01-01  Test
    a  1 A
    b
```

Output for v1.20.1
```
$ hledger -f futurevaluation.journal bal a -N -V
                 1 B  a
$ hledger -f futurevaluation.journal bal a -N --value=end
                 2 B  a
```

Proposed output
```
$ hledger -f futurevaluation.journal bal a -N -V
                 2 B  a
$ hledger -f futurevaluation.journal bal a -N --value=end
                 2 B  a
```

**Comment by Xitian9** [comment]
You can also see here https://github.com/Xitian9/hledger/tree/defaultvaluation for an implementation of this.

**Comment by simonmichael** [comment]
Thanks! And here are the valuation dates chosen above (and for --value=now):
```
$ hledger-1.20.1 -f pastvaluation.j bal a -N --debug -V          |& grep 'valuation date'
valuation date: 2020-12-18
$ hledger-1.20.1 -f pastvaluation.j bal a -N --debug --value=end |& grep 'valuation date'
valuation date: 2000-01-01
$ hledger-1.20.1 -f pastvaluation.j bal a -N --debug --value=now |& grep 'valuation date'
valuation date: 2020-12-18
$ hledger-1.20.1 -f futurevaluation.j bal a -N --debug -V          |& grep 'valuation date'
valuation date: 2020-12-18
$ hledger-1.20.1 -f futurevaluation.j bal a -N --debug --value=end |& grep 'valuation date'
valuation date: 2100-01-01
$ hledger-1.20.1 -f futurevaluation.j bal a -N --debug --value=now |& grep 'valuation date'
valuation date: 2020-12-18
```

**Comment by simonmichael** [comment]
Sorry @Xitian9, I'm having a hard time following this right now. The "fault" is mine, not yours. 

Can we put that on pause and step back to @aragaer's original issue for a moment. They were surprised by the change of result when adding -Y:
```
$ cat b.j
P 2020-11-20 commodity 1

2020-11-20
 (assets)  1 commodity

P 2020-11-22 commodity 2

$ hledger-1.20.1 -f b.j bal --value=end -e 2021 --debug     # picks valuation date: 2020-11-20
                   1  assets
--------------------
                   1
$ hledger-1.20.1 -f b.j bal --value=end -e 2021 --debug -Y  # picks valuation date: 2020-12-31
Period-end value changes in 2020:

        || 2020 
========++======
 assets ||    2 
--------++------
        ||    2 
```

I think there's a clear bug here: the first command should also have picked 2020-12-31 as the valuation date, per docs, because of --value=end and -e 2021.

**Comment by simonmichael** [comment]
And back to what you are talking about:

> I would propose just having -V for single period reports behave identically to --value=end

Ie in that situation, make -V equivalent to --value=end rather than --value=now. 

Well, since in practice P directives are usually in the past, and now that "end" also means "date of last P directive", I think this change would mostly give the same results as before. (I'm not sure if that is consistent with your pastvaluation example above, or not.)

But it would give different results with future-dated things, as in your futurevaluation example. To make things clearer, could you add the proposed valuation dates to your "proposed output" above ? Or, we can just talk about valuation dates instead of showing balance report output, it'll be easier to follow.

**Comment by simonmichael** [comment]
Digression: when showing a valued report, I wonder if it would be good to add a heading communicating the valuation date, to reduce confusion; or would it be too noisy.

**Comment by Xitian9** [comment]
> Sorry @Xitian9, I'm having a hard time following this right now. The "fault" is mine, not yours.
> 
> Can we put that on pause and step back to @aragaer's original issue for a moment. They were surprised by the change of result when adding -Y:
> 
> ```
> $ cat b.j
> P 2020-11-20 commodity 1
> 
> 2020-11-20
>  (assets)  1 commodity
> 
> P 2020-11-22 commodity 2
> 
> $ hledger-1.20.1 -f b.j bal --value=end -e 2021 --debug     # picks valuation date: 2020-11-20
>                    1  assets
> --------------------
>                    1
> $ hledger-1.20.1 -f b.j bal --value=end -e 2021 --debug -Y  # picks valuation date: 2020-12-31
> Period-end value changes in 2020:
> 
>         || 2020 
> ========++======
>  assets ||    2 
> --------++------
>         ||    2 
> ```
> 
> I think there's a clear bug here: the first command should also have picked 2020-12-31 as the valuation date, per docs, because of --value=end and -e 2021.

Yes, this will be resolved when #1418 is merged. The issue is that balance reports will trim trailing and leading empty columns by default when `--empty` is not given, so even though an explicit end is specified, that gets trimmed to the date of the last posting before the explicitly specified end. #1418 means that this will no longer happen.

**Comment by simonmichael** [comment]
#1418 merged.

**Comment by Xitian9** [comment]
> > I would propose just having -V for single period reports behave identically to --value=end
> 
> Ie in that situation, make -V equivalent to --value=end rather than --value=now.
> 
> Well, since in practice P directives are usually in the past, and now that "end" also means "date of last P directive", I think this change would mostly give the same results as before. (I'm not sure if that is consistent with your pastvaluation example above, or not.)

This is correct.

> But it would give different results with future-dated things, as in your futurevaluation example. To make things clearer, could you add the proposed valuation dates to your "proposed output" above ? Or, we can just talk about valuation dates instead of showing balance report output, it'll be easier to follow.

I've added the proposed valuation dates above, and am happy to just talk about those from now on. I've also submitted #1422 with the proposed changes.

**Comment by simonmichael** [comment]
#1422 merged.

**Comment by simonmichael** [comment]
#1418 and #1422 are merged in master, fixing this issue there, but not in 1.20-branch. Cherry picking #1418 there causes 12 valuation test failures. I'm inclined to leave this unfixed in 1.20.3.

**Comment by simonmichael** [comment]
Though #1418 is also the fix for #1396. @Xitian9 if it's easy for you to adapt it for 1.20-branch, without disrupting other things - I'm not certain of all the impacts - it might be nice to have in 1.20.3. If it's any hassle, let's not worry about it.
